
Status: FIONBIO/FIOASYNC is in http://www.freebsd.org/cgi/query-pr.cgi?pr=138526
	/dev/full already rejected in upstream (kern/68961)

--- a/sys/dev/null/null.c
+++ b/sys/dev/null/null.c
@@ -43,13 +43,16 @@
 
 /* For use with destroy_dev(9). */
 static struct cdev *null_dev;
+static struct cdev *full_dev;
 static struct cdev *zero_dev;
 
 static d_write_t null_write;
+static d_write_t full_write;
 static d_ioctl_t null_ioctl;
 static d_read_t zero_read;
 
 #define NULL_MINOR	2
+#define FULL_MINOR	3
 #define ZERO_MINOR	12
 
 static struct cdevsw null_cdevsw = {
@@ -60,6 +63,14 @@
 	.d_name =	"null",
 };
 
+static struct cdevsw full_cdevsw = {
+	.d_version =	D_VERSION,
+	.d_read =	(d_read_t *)nullop,
+	.d_write =	full_write,
+	.d_ioctl =	null_ioctl,
+	.d_name =	"full",
+};
+
 static struct cdevsw zero_cdevsw = {
 	.d_version =	D_VERSION,
 	.d_read =	zero_read,
@@ -81,11 +92,25 @@
 
 /* ARGSUSED */
 static int
+full_write(struct cdev *dev __unused, struct uio *uio, int flags __unused)
+{
+	uio->uio_resid = 0;
+
+	return (ENOSPC);
+}
+
+#include <sys/filio.h>
+/* ARGSUSED */
+static int
 null_ioctl(struct cdev *dev __unused, u_long cmd, caddr_t data __unused,
     int flags __unused, struct thread *td)
 {
 	int error;
 
+	if (cmd == FIONBIO)
+		return 0;
+	if ((cmd == FIOASYNC) && ((*(int *)data) == 0))
+		return 0;
 	if (cmd != DIOCSKERNELDUMP)
 		return (ENOIOCTL);
 	error = priv_check(td, PRIV_SETDUMPER);
@@ -117,12 +142,15 @@
 		zbuf = (void *)malloc(PAGE_SIZE, M_TEMP, M_WAITOK | M_ZERO);
 		null_dev = make_dev(&null_cdevsw, NULL_MINOR, UID_ROOT,
 			GID_WHEEL, 0666, "null");
+		full_dev = make_dev(&full_cdevsw, FULL_MINOR, UID_ROOT,
+			GID_WHEEL, 0666, "full");
 		zero_dev = make_dev(&zero_cdevsw, ZERO_MINOR, UID_ROOT,
 			GID_WHEEL, 0666, "zero");
 		break;
 
 	case MOD_UNLOAD:
 		destroy_dev(null_dev);
+		destroy_dev(full_dev);
 		destroy_dev(zero_dev);
 		free(zbuf, M_TEMP);
 		break;
